### ISSUE TITLE
Broken NVDA Support for Visual Studio 2017.3

### ISSUE THREAD

**Issue by francipvb** [nvaccess/nvda#7504]
Hello,

The NVDA support is broken with the last update for Visual Studio 2017 (15.3).

Some things that are not working:

* Object explorer. I.E. If you expand a node and go to a child NVDA thinks that the focused object is the next node (not the child)
* IntelliSense: Support for IntelliSense is completely broken. Autocompletion, quick actions, Code snippets and Surroundings popups aren't read by NVDA

According to [this post](https://blogs.msdn.microsoft.com/visualstudio/2017/08/14/accessibility-improvements-in-visual-studio-2017-version-15-3/), this update features 1700 accessibility fixes.

However, Narrator works well with all of these features :confused:

Cheers,

**Comment by Neurrone** [comment]
Have you reported this to Microsoft? I'll ping them on twitter with a link to this issue.

**Comment by francipvb** [comment]
Hi,

No, I didn't.

With NVDA 2017.3 reading of intellicense support has been fixed, but object explorer is buggy.

Cheers,

**Comment by francipvb** [comment]
Hello,

I got useful information from the log:

```
IO - speech.speak (12:23:49.693):
Speaking [u'di\xe1logo']
IO - speech.speak (12:23:49.694):
Speaking [u'VsObjectBrowserTypesPane  \xe1rbol']
IO - speech.speak (12:23:49.703):
Speaking [u'Identity  Contra\xeddo']
IO - inputCore.InputManager.executeGesture (12:23:52.421):
Input: kb(desktop):rightArrow
IO - speech.speak (12:23:52.447):
Speaking [u'expandido']
IO - inputCore.InputManager.executeGesture (12:23:53.417):
Input: kb(desktop):downArrow
DEBUGWARNING - IAccessibleHandler.accessibleObjectFromEvent (12:23:53.546):
oleacc.AccessibleObjectFromEvent with window 2884654, objectID 75120 and childID 0: [Error -2147467259] Error no especificado
DEBUGWARNING - IAccessibleHandler.accessibleObjectFromEvent (12:23:53.551):
oleacc.AccessibleObjectFromEvent with window 2884654, objectID 75121 and childID 0: [Error -2147467259] Error no especificado
DEBUGWARNING - IAccessibleHandler.accessibleObjectFromEvent (12:23:53.552):
oleacc.AccessibleObjectFromEvent with window 2884654, objectID 75122 and childID 0: [Error -2147467259] Error no especificado
DEBUGWARNING - IAccessibleHandler.accessibleObjectFromEvent (12:23:53.553):
oleacc.AccessibleObjectFromEvent with window 2884654, objectID 75123 and childID 0: [Error -2147467259] Error no especificado
DEBUGWARNING - IAccessibleHandler.accessibleObjectFromEvent (12:23:53.555):
oleacc.AccessibleObjectFromEvent with window 2884654, objectID 75124 and childID 0: [Error -2147467259] Error no especificado
DEBUGWARNING - IAccessibleHandler.accessibleObjectFromEvent (12:23:53.555):
oleacc.AccessibleObjectFromEvent with window 2884654, objectID 75125 and childID 0: [Error -2147467259] Error no especificado
DEBUGWARNING - IAccessibleHandler.accessibleObjectFromEvent (12:23:53.556):
oleacc.AccessibleObjectFromEvent with window 2884654, objectID 75126 and childID 0: [Error -2147467259] Error no especificado
DEBUGWARNING - IAccessibleHandler.accessibleObjectFromEvent (12:23:53.559):
oleacc.AccessibleObjectFromEvent with window 2884654, objectID 75127 and childID 0: [Error -2147467259] Error no especificado
DEBUGWARNING - IAccessibleHandler.accessibleObjectFromEvent (12:23:53.559):
oleacc.AccessibleObjectFromEvent with window 2884654, objectID 75128 and childID 0: [Error -2147467259] Error no especificado
DEBUGWARNING - IAccessibleHandler.accessibleObjectFromEvent (12:23:53.559):
oleacc.AccessibleObjectFromEvent with window 2884654, objectID 75129 and childID 0: [Error -2147467259] Error no especificado
IO - speech.speak (12:23:53.586):
Speaking [u'Microsoft.AI.DependencyCollector  Contra\xeddo  no seleccionado']
```

Hope this helps.

Cheers,

**Comment by Neurrone** [comment]
@francipvb Could you post that with the language set to English?

**Comment by francipvb** [comment]
Hello,

Here is a complete session log taken from an instance from master branch.

Cheers,
[nvda.txt](https://github.com/nvaccess/nvda/files/1266749/nvda.txt)

**Comment by shadyar** [comment]
Hi
I have the same problem with "Object explorer" on VS 2017. in addition NVDA is some times stopped on "solution explorer". 
NVDA Version: snapshot, next-14660,06727657
Windows: v10 x64
user type: administrator

**Comment by robinettl** [comment]
Any further updates on this issue?  I am using VS 2017 Community and am having trouble with NVDA (latest version) on Windows 10.  IntelliSense works but it reads too much, making it almost useless.  For example, say I've typed a line of code and toward the end of that line an IntelliSense menu pops up to help me complete a command.  NVDA reads that entire line of code followed by whatever item is highlighted in the IntelliSense popup.  This is impractical because I have to wait for that whole line of code to be read aloud every time I move up or down in the IntelliSense menu. Strangely, I have basically the same setup at work, with the exception it's the commercial version of VS and not Community, but same version of NVDA and also Windows 10 and I don't have this issue.  It actually works fine.  I have no idea what the difference might be.  Thanks all.

**Comment by ehollig** [comment]
CC @derekriemer @josephsl

**Comment by robinettl** [comment]
I see that somebody replied but I don't see any actual comment. I am new to this site. Am I missing something?

**Comment by man0528** [comment]
I can't use IntelliSense properly so far， I am helpless

**Comment by leonardder** [comment]
@francipvb: does your issue still apply to the most recent version (which is 2017.6 I believe)?

**Comment by francipvb** [comment]
Hello,

Yes, it does,, unfotunately.

Cheers



Enviado desde mi Samsung Mobile de Claro

**Comment by leonardder** [comment]
@michaeldcurran: As you told me you're more and more using VS 2017 yourself, could you elaborate on what issues you encounter and whether these require external fixes?

**Comment by francipvb** [comment]
Hello people,

I was reviewing this issue today and yesterday.

I found that the object explorer is treated as IAccessible objects rather than WPF objects.

Cheers,

**Comment by leonardder** [comment]
@francipvb: what happens if you do the following from a python console:
`nav.appModule.isGoodUIAWindow=lambda hwnd: True`

**Comment by francipvb** [comment]
Hello,

It magically works correct with this.

Cheers,

**Comment by francipvb** [comment]
The only problem is that NVDA doesn't recognizes expand and contract events (state changes) and in last items parents for children aren't set up correctly, apparently.

See this chunk and see for yourself:

````

IO - speech.speak (12:37:23.022):
Speaking [u'VsObjectBrowserTypesPane  tree view']
IO - speech.speak (12:37:23.032):
Speaking [u'level 1  WindowsBase  collapsed']
IO - inputCore.InputManager.executeGesture (12:37:23.463):
Input: kb(desktop):end
IO - inputCore.InputManager.executeGesture (12:37:24.010):
Input: kb(desktop):upArrow
IO - speech.speak (12:37:24.226):
Speaking [u'System.Xml.Linq  collapsed  level 1']
IO - inputCore.InputManager.executeGesture (12:37:24.605):
Input: kb(desktop):downArrow
IO - speech.speak (12:37:24.801):
Speaking [u'WindowsBase  collapsed  level 1']
IO - inputCore.InputManager.executeGesture (12:37:25.898):
Input: kb(desktop):rightArrow
IO - inputCore.InputManager.executeGesture (12:37:26.769):
Input: kb(desktop):downArrow
IO - speech.speak (12:37:26.993):
Speaking [u'MS.Internal  collapsed  level 1']
IO - inputCore.InputManager.executeGesture (12:37:28.421):
Input: kb(desktop):upArrow
IO - speech.speak (12:37:28.647):
Speaking [u'WindowsBase  expanded  level 1']
IO - inputCore.InputManager.executeGesture (12:37:29.983):
Input: kb(desktop):leftArrow
IO - inputCore.InputManager.executeGesture (12:37:31.496):
Input: kb(desktop):upArrow
IO - speech.speak (12:37:31.727):
Speaking [u'System.Xml.Linq  collapsed  level 1']
IO - inputCore.InputManager.executeGesture (12:37:32.026):
Input: kb(desktop):rightArrow
IO - inputCore.InputManager.executeGesture (12:37:32.598):
Input: kb(desktop):downArrow
IO - speech.speak (12:37:32.826):
Speaking [u'level 2  MS.Internal.Xml.Linq.ComponentModel  collapsed']
IO - inputCore.InputManager.executeGesture (12:37:33.578):
Input: kb(desktop):upArrow
IO - speech.speak (12:37:33.789):
Speaking [u'level 1  System.Xml.Linq  expanded']
IO - inputCore.InputManager.executeGesture (12:37:34.019):
Input: kb(desktop):leftArrow
IO - inputCore.InputManager.executeGesture (12:37:34.522):
Input: kb(desktop):rightArrow
IO - inputCore.InputManager.executeGesture (12:37:34.809):
Input: kb(desktop):downArrow
IO - speech.speak (12:37:35.035):
Speaking [u'level 2  MS.Internal.Xml.Linq.ComponentModel  collapsed']
IO - inputCore.InputManager.executeGesture (12:37:35.122):
Input: kb(desktop):downArrow
IO - speech.speak (12:37:35.335):
Speaking [u'level 1  System.Xml.Linq  collapsed']
IO - inputCore.InputManager.executeGesture (12:37:35.834):
Input: kb(desktop):upArrow
IO - speech.speak (12:37:36.092):
Speaking [u'level 2  MS.Internal.Xml.Linq.ComponentModel  collapsed']
IO - inputCore.InputManager.executeGesture (12:37:36.134):
Input: kb(desktop):upArrow
IO - speech.speak (12:37:36.354):
Speaking [u'level 1  System.Xml.Linq  expanded']
IO - inputCore.InputManager.executeGesture (12:37:36.585):
Input: kb(desktop):leftArrow
IO - inputCore.InputManager.executeGesture (12:37:38.970):
Input: kb(desktop):downArrow
IO - speech.speak (12:37:39.157):
Speaking [u'WindowsBase  collapsed  level 1']
```

However, neither narrator doesn't work correctly in this situation.

Cheers,

**Comment by leonardder** [comment]
There's something strange happening here which I, by coincidence, also saw on another system in a WPF app. For the WPF app in question, IAccessible was used for native UIA controls as well. Further inspection showed that UIAutomationCore.UiaHasServerSideProvider returns False and the UIA provider is ineed the MSAA proxy. Yet, the UIA implementation is much more usable.
cc @josephsl

**Comment by francipvb** [comment]
I will link your comment to an issue I opened in the visual studio support.

Cheers,

**Comment by robinettl** [comment]
I reported a very similar issue for the object browser  in Microsoft SQL server management studio. When I get off work I will find the number and comment it here, as the fix for one might mean a fix for the other

**Comment by francipvb** [comment]
Hello again,

I will open separate issues for this.

Cheers,

**Comment by francipvb** [comment]
Please check #9311 and #9313 

Cheers,

**Comment by leonardder** [comment]
@francipvb HOw is this in Visual Studio 2019?

**Comment by francipvb** [comment]
I don't know, the Visual Studio AppModule have a dirty hack for object explorer.

Sorry about Intellisense, I have a fix working very well and very usable, however, there is an issue with UIA events from NVDA side, they are processed too late.

**Comment by francipvb** [comment]
Hello again @leonardder 

The object explorer annoying bug (without the dirty fix) is still there.

Regarding Intellisense, I found that the best solution is to intercept the liveRegionChanged event from the appModule and just stop speech before the call to nextHandler.

The issue is that this event is fired various times during the loading process of the Intellisense popup list, so I have to find a workaround this before I can post this.

Cheers,

**Comment by leonardder** [comment]
Could you elaborate on why you think NVDA processes events too late? How 
can that be reproduced?

**Comment by francipvb** [comment]
I'm not sure how I can reproduce this reliably, but it occurs frequently, particularly when first loading Visual Studio Intellisense.

Maybe a try with an WPF example with simulated high workload...

I've found that in Visual Studio 2017, for example, the start page loads but you have to wait until the user interface loads completely to get focus events at appropriate time.

Sorry, but this is a bit tricky, at least for me.

**Comment by Luke-Robinett** [comment]
FYI none of this has been fixed and it's two years later now. Are we just going to endlessly swap error logs and play guessing games on the issues page or is someone at NVDA going to actually fix these problems? 1. IntelliSense messages about parameters and functions are not spoken at all. 2. the menus that list members when you type a period in your code do not work properly. Instead of NVDA just reading the selected menu item as you scroll up or down, it reads the entire line of code followed by that menu option. Narrator does not exibit these behaviors. As someone who donates on a recurring monthly basis, I'd really like to know what the delay here is and why someone hasn't stepped up to the plate. NVAccess is no longer just some small scrappy basement software operation - they have thousands of donors big and small and are a widely used screen reader platform. They need to get their technical support up to speed. My work upgraded to VS 2017 so I'm stuck with this nearly unusable behavior from NVDA until you fix it.

**Comment by josephsl** [comment]
Hi Luke,

First, you are right - we (NVDA people) should have stepped up and kept an eye on this issue. The reason why we weren't able to do so is because of different forces pulling us into different directions - Python 3, speech refactor, vision provider framework and others.

Second, there is someone working on an NVDA add-on for Visual Studio, and I heard this issue is resolved (we need to double-check that). Hopefully the fix from the add-on can be incorporated into NVDA in the future.

I know that what I wrote above may not be the most satisfying answer you are seeking, but note that developers are looking into ways to resolve issues you've reported.

Thanks.

**Comment by francipvb** [comment]
Hello,

> 
> 
> FYI none of this has been fixed and it's two years later now. Are we just going to endlessly swap error logs and play guessing games on the issues page or is someone at NVDA going to actually fix these problems? 1. IntelliSense messages about parameters and functions are not spoken at all. 2. the menus that list members when you type a period in your code do not work properly. Instead of NVDA just reading the selected menu item as you scroll up or down, it reads the entire line of code followed by that menu option. Narrator does not exhibit these behaviors. Please fix this asap. Too much talking and not enough fixing. Two years is ridiculous.
Sorry about that. We have a solution for this. However, the solution was rejected because it doesn't applies to SSMS.

If I manage to have SSMS installed this afternoon I will post a PR to solve this definitely.

Cheers,

**Comment by francipvb** [comment]
Hello @Luke-Robinett,

Please try this [package](https://ci.appveyor.com/api/buildjobs/r3eysnxg1ug9f66o/artifacts/VisualStudioExtended-master-12.nvda-addon).

Cheers,

**Comment by Luke-Robinett** [comment]
Thanks folks. I apologize for my terse comment. I've removed it as I felt it wasn't professional nor the impression I want to put out there. I understand things take time and that everyone believes his or her issue is the most important and should deserve most of the attention. As a developer myself, I realize not everything has a convenient solution, but it's great to hear that progress is being made. If I knew Python I'd jump in and see if I could help resolve the issue but I'm not there yet. Thanks again.

**Comment by leonardder** [comment]
@francipvb Any luck with SSMS so far? 

Cc @sjfbol

**Comment by leonardder** [comment]
Is any of you still using Visual Studio 2017 now version 2019 is pretty mature?

**Comment by man0528** [comment]
I am using VS2019, but NVDA's performance on the code editor is still not ideal
IntelliSense is basically unavailable, The method prompt list will have redundant content

发件人: Leonard de Ruijter<mailto:notifications@github.com>
发送时间: 2020年7月25日 15:53
收件人: nvaccess/nvda<mailto:nvda@noreply.github.com>
抄送: man0528<mailto:manchen_0528@outlook.com>; Comment<mailto:comment@noreply.github.com>
主题: Re: [nvaccess/nvda] Broken NVDA Support for Visual Studio 2017.3 (#7504)


Is any of you still using Visual Studio 2017 now version 2019 is pretty mature?

―

**Comment by leonardder** [comment]
Have you tried NVDA Alpha snapshots with the "Enable selective registration for UI Automation events and property changes" option enabled in NVDA"s advanced settings? THat should improve things dramatically in Visual Studio.

**Comment by man0528** [comment]
Hello!
as you said, This can indeed significantly improve performance, But I still want to know how to read IntelliSense lists more efficiently.
eg, input "Console.". Then press the Down key to initially speak "Console" before each item, And how to hear the parameter prompt in the method.
Thanks.


________________________________
发件人: Leonard de Ruijter <notifications@github.com>
发送时间: Saturday, July 25, 2020 4:05:05 PM
收件人: nvaccess/nvda <nvda@noreply.github.com>
抄送: man0528 <manchen_0528@outlook.com>; Comment <comment@noreply.github.com>
主题: Re: [nvaccess/nvda] Broken NVDA Support for Visual Studio 2017.3 (#7504)


Have you tried NVDA Alpha snapshots with the "Enable selective registration for UI Automation events and property changes" option enabled in NVDA"s advanced settings? THat should improve things dramatically in Visual Studio.

―

**Comment by cary-rowen** [comment]
Is anyone paying attention to this issue?
The same problem is still not fixed on 2020.3, I am really upset.

**Comment by robinettl** [comment]
I wouldn’t hold your breath, unfortunately. I went back-and-forth at length with developers in another comment thread and ultimately was told I was out of line for voicing what I feel is understandable frustration. on one hand, I get that NVDA is a small team compared to the likes of Freedom Scientific, Microsoft, Apple etc. But at the same time, from what I’ve read, NVDA has pretty significant market share nowadays and I must imagine they are getting plenty of sizable donations from individuals and organizations alike. It seems to me they are leaning pretty heavily on that excuse of just being the little guys and not having the resources to make their products truly professional. I continue to donate on a recurring basis to them. I expect the same level of quality I would from JAWS, voiceover or other commercial screen reader products. Quite honestly, I’m getting tired of the excuses. I’ll leave it at that for now.

**Comment by leonardder** [comment]
@cary-rowen wrote:

> Is anyone paying attention to this issue?
> The same problem is still not fixed on 2020.3, I am really upset.

There are several issues mentioned in this issue, which makes it pretty hard to distilate the relevant parts from it. May be you could explain in a [new issue](https://github.com/nvaccess/nvda/issues/new?template=bug_report.md) we're you are suffering from? Also note that after 2020.3 was released, new alpha builds were created that should improve the overall behavior of IntelliSense.

@robinettl wrote:

> I get that NVDA is a small team compared to the likes of Freedom Scientific, Microsoft, Apple etc. 

This is correct. Also note that several people, like me, are only volunteers. I can only contribute to NVDA in my spare time, which is sometimes somewhat scarse with a full time job and growing children.

> I expect the same level of quality I would from JAWS, voiceover or other commercial screen reader products.

Well, in that case, I have to say your expectations are not right. You simply can't expect the same from an open source product as from a commercial product. Usually, paying for a commercial product gives you support right, ability to upgrade to newer versions, expect that bugs are fixed in a timely fashion. Donating for an open source product doesn't give you the same rights for that product. Of course you are free to switch to a commercial screen reader if that fits you better.

**Comment by leonardder** [comment]
I have briefly discussed this with @feerrenrut, and we both agree that it is best to close this issue. There are several reasons for this.

1. Strictly spoken, the issue is targetting Visual Studio 2017 whereas 2019 is now the mainstream standard to use.
2. While a typical bug report is meant to target only one single issue, this thread contains several of them which makes it pretty hard to follow what is actually still relevant and what not. This is likely the reason why the issue has undergone some frustration as well.

I'd like to invite everyone who suffers from Visual Studio issues to do the following:

1. Find out whether there's already an issue for it. A way to do this is looking in the list with [Visual Studio specific issues](https://github.com/nvaccess/nvda/labels/app%2Fvisual-studio)
2. If not already done, enable the "Enable selective registration for UI Automation events and property changes" setting in NVDA's advanced configuration.
3. Test whether the problem still exsits in [NVDA development snapshots](https://www.nvaccess.org/files/nvda/snapshots/)
4. If not, open a new issue, filling out the bug template as requested. Please try to limit your thread to a single issue. Saying that parameter hints aren't read is unrelated to Visual Studio performance, for example.

As I am a daily Visual Studio user, I might be able to reproduce some of the issues instantly and provide an easy fix, but given my pretty busy schedule, I can't promise anything.

Thanks for the understanding.
